File: timetravel.py
import re
import os
import sys
import random
import json
from datetime import datetime, date, time as _time, timedelta
from functools import wraps


IGNORE = os.path.dirname(sys.argv[0])


HTML_TEMPLATE = """
<html>
<head>
    <link rel="stylesheet" href="https://code.jquery.com/ui/1.11.2/themes/smoothness/jquery-ui.css">
    <script src="https://code.jquery.com/jquery-latest.js"></script>
    <script src="https://code.jquery.com/ui/1.11.2/jquery-ui.js"></script>
    <script src="timetravel.js"></script>
    <style>
        div.scroll {
            height: 450px;
            overflow-y: auto;
        }
        .big {
            width: 200px;
            vertical-align: top;
        }
    </style>
    <script type='text/javascript'>
        window.onload = function() {
            $("button#far-left").click(far_left_click);
            $("button#left").click(left_click);
            $("button#right").click(right_click);
            $("button#far-right").click(far_right_click);
            $.ajax({
                url: 'JSONFILE',
                type: 'GET',
                cache: false,
                dataType: 'json'
            }).fail(function(e) {
                alert("Bad JSON file");
                console.log(e);
            }).done(function(data) {
                console.log("D");
                $.map(data.filenames, function(j) {
                    prepare_file(j);
                });
                data = data.data;
                steps = data;
                num_steps_to_process = data.length - 1;
            });
        }
    </script>
</head>
<body>
    <h1 id="filename"></h1>
    <div>
        <button id="far-left">&lt;&lt;&lt;</button>
        <button id="left">&lt;</button>
        <button id="right">&gt;</button>
        <button id="far-right">&gt;&gt;&gt;</button>
    </div>
    <div>
        <table>
            <tr>
                <td class="big">
                    <table>
                        <tr><td><div>
                            <div id="target" class="scroll">
                            </div>
                        </div></td></tr>
                    </table>
                </td>
                <td class="big">
                    <pre id="vars" class="scroll">
                    </pre>
                    <div id="returnvalue">
                    </div>
                </td>
            </tr>
        </table>
    </div>
</body>
</html>
"""


def time_travel(htmlfile=None):
    LOGFILE = 'x.log'
    f_index = [0]
    forward = {}    # map filename to index
    fromframe = {}  # map id(frame) to index

    jsonfile = 'X{0:x}.json'.format(random.randint(1, 1<<32))
    if htmlfile is None:
        htmlfile = 'X{0:x}.html'.format(random.randint(1, 1<<32))
        print 'HTML file for time travel debug is {0}'.format(htmlfile)

    def make_json_file(logfile, jsonfile):
        outf = open(jsonfile, 'w')
        outf.write('{"data":[')
        r = re.compile("^([^ ]+) ([^ ]+) ([^ ]+) ([^ ]+) (.*)")
        filenames = []
        mode = 0
        started = False
        for line in open(logfile).readlines():
            line = line.rstrip()
            if line == '====':
                mode = 1
                outf.write('],"filenames":')
            elif mode == 0:
                groups = r.match(line).groups()
                if groups[0] in ('call', 'line', 'return'):
                    d = {
                        'F': groups[1],
                        'L': int(groups[2]),
                        'V': groups[4]
                    }
                    if groups[0] == 'return':
                        d['E'] = groups[0]
                        d['A'] = groups[3]
                    if started:
                        outf.write(',')
                    outf.write(json.dumps(d) + '\n')
                    started = True
            else:
                filenames.append(line)
        outf.write(json.dumps(filenames) + '}')
        outf.close()
        os.unlink(logfile)

    def fix_json_char(ch):
        # json.dumps barfs on ASCII codes outside vanilla EN-US range
        # and also on datetime.* types
        if 32 <= ord(ch) <= 126:
            return str(ch)
        else:
            return ' '

    def make_json_dumpable(obj):
        if obj is None or isinstance(obj, (int, long, float)):
            return obj
        elif isinstance(obj, complex):
            return {'real': obj.real, 'imag': obj.imag}
        elif isinstance(obj, (str, unicode)):
            return ''.join((fix_json_char(x) for x in obj))
        elif isinstance(obj, (set, list, tuple)):
            return [make_json_dumpable(x) for x in obj]
        elif isinstance(obj, datetime) or isinstance(obj, date) or isinstance(obj, _time):
            return obj.isoformat()
        elif isinstance(obj, timedelta):
            return str(obj)
        elif isinstance(obj, (set, list, tuple)):
            return [make_json_dumpable(x) for x in obj]
        elif isinstance(obj, dict):
            return {
                make_json_dumpable(k):
                make_json_dumpable(v)
                for k, v in obj.items()
            }
        else:
            try:
                return repr(obj)
            except:
                return "???"
    def decorator(f):
        outf = open(LOGFILE, 'w')

        def trace_func(frame, event, arg):
            filename = os.path.realpath(frame.f_code.co_filename)
            if filename.startswith("/usr"):
                return trace_func
            i = id(frame)
            fi = fromframe.get(i, None)
            if fi is None:
                # filename = os.path.realpath(frame.f_code.co_filename)
                filename = frame.f_code.co_filename.replace(IGNORE, "")
                fi = forward.get(filename, None)
                if fi is None:
                    fi = f_index[0]
                    f_index[0] += 1
                    forward[filename] = fi
                fromframe[i] = fi
            outf.write(
                '{0} {1} {2} {3} '.format(event, fi, frame.f_lineno, arg) +
                json.dumps(make_json_dumpable(frame.f_locals)) +
                '\n'
            )
            if event == 'return':
                fromframe[i] = None
            return trace_func

        @wraps(f)
        def wrapped(*args, **kw):
            orig = sys.gettrace()
            try:
                sys.settrace(trace_func)
                return f(*args, **kw)
            finally:
                outf.write("====\n")
                items = [(v, k) for k, v in forward.items()]
                items.sort()
                for k, v in items:
                    outf.write("{0}\n".format(v))
                outf.close()
                sys.settrace(orig)
                make_json_file(LOGFILE, jsonfile)
                open(htmlfile, 'w').write(HTML_TEMPLATE.replace("JSONFILE", jsonfile))
        return wrapped
    return decorator

File: timetravel.js
const pythonVariableName = /[a-zA-Z][a-zA-Z0-9]*/g;
let steps = [];
let which_step = 0;
let my_tables = {};
let most_recent_url = "";
let num_steps_to_process = 0;
let prefix = null;
let files = [];

function prepare_file(url) {
    if (my_tables[url]) {
        return;
    }
    if (prefix === null) {
        let before = "";
        let after = window.location.href;
        for (let i = 0; i < 3; i++) {
            let n = after.indexOf("/");
            before += after.substr(0, n + 1);
            after = after.substr(n + 1);
        }
        //prefix = before.substr(0, before.length - 1);
        prefix = before;
    }
    files.push(url);
    $.get(prefix + url).done(function(data) {
        let table = $('<table>');
        let linenum = 1;
        $.map(data.split("\n"), function (line) {
            let row = $('<tr id="L' + linenum + '">');
            let num = "" + linenum;
            while (num.length < 6) {
                num = " " + num;
            }
            line = num + "  " + line;
            let td = $('<td><code style="white-space: pre">'
                + line
                + '</code></td>');
            row.append(td);
            table.append(row);
            linenum += 1;
        });
        my_tables[url] = table;
	display_step();
    });
}

function display_file(url, vars, event, arg) {
    vars = JSON.parse(vars);
    if (url != most_recent_url) {
        most_recent_url = url;
        $('#filename').html('<h3>' + url + '</h3>');
        const target = $('div#target');
        target.empty();
        target.append(my_tables[url]);
    }
    let table = $('<table>')
        .attr('border', '1px solid black')
        .attr('border-collapse', 'collapse');
    for (key in vars) {
        const value = vars[key];
        table.append($(
            '<tr><td>' + key + '</td><td>'
            + value + '</td></tr>'));
    }
    $("#vars").empty();
    $("#vars").append(table);
    $("#returnvalue").empty();
    if (event == 'return') {
        let x = 'Return <tt>' + arg + '</tt>';
        $("#returnvalue").html(x);
    }
}
function unmark_line(selector) {
    $(selector).css("background-color", "");
}
function mark_line(selector) {
    const x = $(selector)[0];
    if (x != undefined) {
        x.scrollIntoView({block: "center"});
    }
    $(selector).css("background-color", "FF0");
}

function display_step() {
    let step = steps[which_step];
    display_file(files[parseInt(step.F)], step.V, step.E, step.A);
    mark_line("#L" + step.L);
}
function far_left_click() {
    left_click_inner(100);
}
function left_click() {
    left_click_inner(1);
}
function left_click_inner(n) {
    if (n === undefined)
        n = 1;
    let step1 = steps[which_step];
    unmark_line("#L" + step1.L);
    while (n > 0) {
        which_step--;
        if (which_step < 0) {
            which_step = 0;
            break;
        }
        n--;
    }
    display_step();
}
function right_click_inner(n) {
    let step1 = steps[which_step];
    while (n > 0) {
        which_step++;
        if (which_step > steps.length - 1) {
            which_step = steps.length - 1;
            break;
        }
        n--;
    }
    unmark_line("#L" + step1.L);
    display_step();
}
function right_click() {
    right_click_inner(1);
}
function far_right_click() {
    right_click_inner(100);
}
